conda/exceptions.py

```diff
@@ -584,177 +584,243 @@ def __init__(self, invalid_spec):
+        return context.remote_connect_timeout_secs, context.remote_read_timeout_secs
+
+    @property
+    def user_agent(self):
+        from .base.context import context
+        return context.user_agent
+
+    @property
+    def error_upload_url(self):
+        from .base.context import context
+        return context.error_upload_url
+
+    def handle_exception(self, exc_type, exc_val, exc_tb):
+        return_code = getattr(exc_val, 'return_code', None)
+        if return_code == 0:
+            return 0
+        elif isinstance(exc_val, CondaError):
+            return self.handle_application_exception(exc_type, exc_val, exc_tb)
+        elif isinstance(exc_val, KeyboardInterrupt):
+            self._print_conda_exception(KeyboardInterrupt, CondaError("KeyboardInterrupt"),
+                                        format_exc())
+            return 1
+        else:
+            return self.handle_unexpected_exception(exc_type, exc_val, exc_tb)
+
+    def handle_application_exception(self, exc_type, exc_val, exc_tb):
+        self._print_conda_exception(exc_type, exc_val, exc_tb)
+        rc = getattr(exc_val, 'return_code', None)
+        return rc if rc is not None else 1
+
+    def _print_conda_exception(self, exc_type, exc_val, exc_tb):
+        print_conda_exception(exc_val, exc_tb)
+
+    def handle_unexpected_exception(self, exc_type, exc_val, exc_tb):
+        error_report = self.get_error_report(exc_type, exc_val, exc_tb)
+        self.print_error_report(error_report)
+        ask_for_upload, do_upload = self._calculate_ask_do_upload()
+        do_upload, ask_response = self.ask_for_upload() if ask_for_upload else do_upload, None
+        if do_upload:
+            self._execute_upload(error_report)
+        self.print_upload_confirm(do_upload, ask_for_upload, ask_response)
+        rc = getattr(exc_val, 'return_code', None)
+        return rc if rc is not None else 1
+
+    def get_error_report(self, exc_type, exc_val, exc_tb):
+        command = ' '.join(ensure_text_type(s) for s in sys.argv)
+        info_dict = {}
+        if ' info' not in command:
             try:
-                message_builder.append(get_main_info_str(error_report['conda_info']))
-            except Exception as e:
-                message_builder.append('conda info could not be constructed.')
-                message_builder.append('%r' % e)
-        message_builder.append('')
+                from .cli.main_info import get_info_dict
+                info_dict = get_info_dict()
+            except Exception as info_e:
+                info_traceback = format_exc()
+                info_dict = {
+                    'error': repr(info_e),
+                    'error_type': info_e.__class__.__name__,
+                    'traceback': info_traceback,
+                }
+
+        error_report = {
+            'error': repr(exc_val),
+            'error_type': exc_type,
+            'command': command,
+            'traceback': exc_tb,
```

Comment:
```diff
-            'traceback': exc_tb,
+            'traceback': ''.join(format_exception(exc_type, exc_val, exc_tb))
```
If `exc_tb` is possibly no `traceback` instance here (`None`), this would have to be handled here, for Python 3.4 compatibility, with something like:
```python
if exc_tb:
    traceback = ''.join(format_exception(exc_type, exc_val, exc_tb))
else:  # Python 3.4 compatibility
    traceback = ''.join(format_exception_only(exc_type, exc_val))
```